File: programs/final.cpp
// copy benchmark: repeatedly copies a 31-byte string into a 32-byte stack
// buffer with strcpy, the canonical fortifiable libc call.
//
// The input FITS the buffer: this benchmark measures the cost of the
// mitigations (canary setup/check, __strcpy_chk bounds check, sanitizer
// instrumentation) when NO attack fires, which is the cost paid on every
// execution in production. Deliberately-overflowing runs are a separate
// experiment, not this one.
//
// The destination is a plain fixed-size array so __builtin_object_size is
// known and _FORTIFY_SOURCE can instrument the call. The source pointer is
// read through a volatile so the copy cannot be hoisted out of the loop, and
// the asm barrier forces buffer to be treated as modified on every iteration.
#include <cstdio>
#include <cstring>

static const char *const TEXT = "ABCDEFGHIJKLMNOPQRSTUVWXYZ01234"; // 31 chars + NUL

int main() {
    constexpr long ITERS = 10000000;

    const char *volatile src = TEXT; // opaque source: strcpy must really run
    char buffer[32];                 // known-size destination: fortifiable

    long long total = 0;
    for (long i = 0; i < ITERS; ++i) {
        std::strcpy(buffer, src);
        asm volatile("" : : "g"(buffer) : "memory");
        total += std::strlen(buffer);
    }

    std::printf("%lld\n", total); // expected: 310000000
    return 0;
}

File: programs/final.rs
// copy benchmark: repeatedly copies a 31-byte string into a 32-byte stack
// buffer. Must stay semantically identical to final.cpp: same buffer size,
// same input, same iteration count, input fits the buffer.
//
// copy_from_slice is the idiomatic bounds-checked equivalent of the C strcpy
// (it panics if the lengths do not match, which is the Rust counterpart of
// the C/C++ mitigations firing). black_box keeps the source opaque and the
// buffer live so the loop cannot be optimized away.
use std::hint::black_box;

const TEXT: &str = "ABCDEFGHIJKLMNOPQRSTUVWXYZ01234"; // 31 bytes

fn main() {
    const ITERS: usize = 10_000_000;

    let mut buffer = [0u8; 32];

    let mut total: i64 = 0;
    for _ in 0..ITERS {
        let src = black_box(TEXT.as_bytes()); // opaque source
        buffer[..src.len()].copy_from_slice(src); // bounds-checked copy
        black_box(&buffer);
        // strlen equivalent: position of the first NUL byte.
        total += buffer.iter().position(|&b| b == 0).unwrap_or(32) as i64;
    }

    println!("{}", total); // expected: 310000000
}
